### PR TITLE
[FIXED] subject renaming with hand crafted reply subject

### PR DESCRIPTION
**Pull request by matthiashanel**
Signed-off-by: Matthias Hanel <mh@synadia.com>

Please read comment in client.go. probably should make this illegal, @ or not.

### CODE REVIEW

File: server/client.go [hunk 1]
```diff
@@ -3493,8 +3493,15 @@ func isReservedReply(reply []byte) bool {
 	if isServiceReply(reply) {
 		return true
 	}
+	rLen := len(reply)
 	// Faster to check with string([:]) than byte-by-byte
-	if len(reply) > gwReplyPrefixLen && string(reply[:gwReplyPrefixLen]) == gwReplyPrefix {
+	if rLen > jsAckPreLen+1 && string(reply[:jsAckPreLen]) == jsAckPre {
+		// Not sure what should happen if a JS message is relayed by a client (probably shouldn't support that)
+		// But it should never contain an @ (used for renaming)
+		if strings.Contains(string(reply[jsAckPreLen:]), "@") {
```
Comment (derekcollison) [inline server/client.go:3501]: Just use bytes.Contains imo.

File: server/client.go [hunk 2]
```diff
@@ -3493,8 +3493,15 @@ func isReservedReply(reply []byte) bool {
 	if isServiceReply(reply) {
 		return true
 	}
+	rLen := len(reply)
 	// Faster to check with string([:]) than byte-by-byte
-	if len(reply) > gwReplyPrefixLen && string(reply[:gwReplyPrefixLen]) == gwReplyPrefix {
+	if rLen > jsAckPreLen+1 && string(reply[:jsAckPreLen]) == jsAckPre {
```
Comment (derekcollison) [inline server/client.go:3498]: We should not allow clients to use that prefix at all.

Comment (derekcollison) [reply server/client.go:3498]: And need to make sure its API prefix aware, etc. Not sure if that is applicable.

**Review by derekcollison — CHANGES_REQUESTED** [review]

File: server/client.go (hunk 2 above)
Comment (matthiashanel) [reply server/client.go:3498]: not applicable

File: server/client.go (hunk 1 above)
Comment (matthiashanel) [reply server/client.go:3501]: I'll deny the prefix, that covers this as well.

Comment (kozlovic) [reply server/client.go]: Do we care if this is under `$JS.ACK.`? I mean maybe we just check that there is no `@` allowed regardless of prefix?

**Review by kozlovic — APPROVED** [review]
LGTM, but see comment about checking of $JS.ACK prefix

**Review by kozlovic — APPROVED** [review]
LGTM (disregard my previous comment since was done was you were updating the PR)

**Comment by matthiashanel** [comment]
tbh, I don't think anyone should be publishing with an ack prefix, `@` or not. This covers the issue as `@` renaming is a JS ack functionality. 

In addition to that I can make `@` an illegal/reserved character. 
But that should probably be done for pub subjects as well. 
Meaning, if `foo@bar` is an illegal reply subject, it should be illegal to publish on that as well.

File: server/client.go (hunk 1 above)
Comment (matthiashanel) [reply server/client.go]: we only look for an `@` if the reply subject has the `jsAckPre`

```
	var remapped bool
	if len(creply) > 0 &&
		c.kind != CLIENT && c.kind != SYSTEM && c.kind != JETSTREAM && c.kind != ACCOUNT &&
		bytes.HasPrefix(creply, []byte(jsAckPre)) {
		// We need to rewrite the subject and the reply.
		if li := bytes.LastIndex(creply, []byte("@")); li != -1 && li < len(creply)-1 {
			remapped = true
			subj, creply = creply[li+1:], creply[:li]
		}
	}
```

Comment (kozlovic) [reply server/client.go]: Yes, I wrote that comment before you updated that we would reject JSACK regardless of presence of '@'. So I am ok with current code.

**Pull request merged**